cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java

```diff
@@ -0,0 +1,288 @@
+import co.cask.tephra.TransactionExecutorFactory;
+import co.cask.tephra.runtime.TransactionClientModule;
+import co.cask.tephra.runtime.TransactionModules;
+import com.google.common.collect.Multimap;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.AbstractModule;
+import com.google.inject.Guice;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Singleton;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.twill.zookeeper.ZKClientService;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Debugging tool for queues in hbase.
+ */
+public class HBaseQueueDebugger extends AbstractIdleService {
+
+  private final HBaseQueueAdmin queueAdmin;
+  private final ZKClientService zkClientService;
+  private final HBaseQueueClientFactory queueClientFactory;
+  private final TransactionExecutorFactory txExecutorFactory;
+
+  @Inject
+  public HBaseQueueDebugger(HBaseQueueAdmin queueAdmin,
+                            HBaseQueueClientFactory queueClientFactory,
+                            ZKClientService zkClientService,
+                            TransactionExecutorFactory txExecutorFactory) {
+    this.queueAdmin = queueAdmin;
+    this.queueClientFactory = queueClientFactory;
+    this.zkClientService = zkClientService;
+    this.txExecutorFactory = txExecutorFactory;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    zkClientService.startAndWait();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    zkClientService.stopAndWait();
+  }
+
+  /**
+   * Only works for {@link co.cask.cdap.data2.transaction.queue.hbase.ShardedHBaseQueueStrategy}.
+   */
+  private void scanQueue(final QueueName queueName) throws Exception {
+    HBaseConsumerStateStore stateStore = queueAdmin.getConsumerStateStore(queueName);
+
+    TransactionExecutor txExecutor = Transactions.createTransactionExecutor(txExecutorFactory, stateStore);
+    txExecutor.execute(new TransactionExecutor.Procedure<HBaseConsumerStateStore>() {
+      @Override
+      public void apply(HBaseConsumerStateStore stateStore) throws Exception {
+        Multimap<Long, QueueBarrier> barriers = stateStore.getAllBarriers();
+        System.out.printf("Got %d barriers", barriers.size()).println();
+
+        QueueStatistics stats = new QueueStatistics();
+        int currentSection = 1;
+
+        for (Long groupId : barriers.keySet()) {
+          System.out.printf("Scanning barriers for group %s", groupId).println();
+          Collection<QueueBarrier> groupBarriers = barriers.get(groupId);
+          Iterator<QueueBarrier> barrierIterator = groupBarriers.iterator();
+
+          QueueBarrier previous;
+          QueueBarrier current = null;
+          while (barrierIterator.hasNext()) {
```

Comment:
Refer to my comment in https://github.com/caskdata/cdap/pull/2290#discussion_r29283524